consensus/tests/checkpoint.rs

```diff
@@ -43,48 +47,194 @@ mod tests {
         );
     }
 
+    fn gen_num_blocks_in_epochs(slot_fill_rate: f64, slot_fill_rate_delta: f64, n: f64) -> i32 {
+        let protocol_constants = ConsensusConstants::default();
+        let nums = gen_num_blocks_in_slots(
+            slot_fill_rate,
+            slot_fill_rate_delta,
+            n * protocol_constants.slots_per_epoch.0 as f64,
+        );
+        nums
+    }
+
+    fn gen_spot_root_epoch_position(slot_fill_rate: f64, slot_fill_rate_delta: f64) -> (u32, u32) {
+        //  We need to simulate both the staking epoch and the next staking epoch,
+        //  the root epoch is the staking epoch. The root epoch position this function generates
+        //   is the epoch number of the staking epoch and the block height the
+        //  staking epoch starts at (the simulation of all blocks preceeding the
+        //  staking epoch
+        let root_epoch_int = 50;
+        let root_block_height =
+            gen_num_blocks_in_epochs(slot_fill_rate, slot_fill_rate_delta, root_epoch_int as f64);
+        return (root_epoch_int, root_block_height as u32);
+    }
+
+    fn convert(x: f64) -> i32 {
+        x.round().rem_euclid(TRANS) as u32 as i32
+    }
+
+    fn gen_num_blocks_in_slots(slot_fill_rate: f64, slot_fill_rate_delta: f64, n: f64) -> i32 {
+        let min_blocks = n * f64::max(slot_fill_rate - slot_fill_rate_delta, 0.0);
+        let max_blocks = n * f64::min(slot_fill_rate + slot_fill_rate_delta, 1.0);
+        let num_blocks_in_slots = (min_blocks + max_blocks) / 2.0;
+        let num = convert(num_blocks_in_slots);
+        return num;
+    }
+
+    fn gen_spot_pair_common_checkpoints(
+        a: &mut ProtocolState,
+        b: &mut ProtocolState,
+        min_a_curr_epoch_slot: u32,
+    ) {
+        // New pairs of spot blocks that share common checkpoints.
+        // The overlap of the checkpoints and the root epoch positions of the blocks
+        // that are generated can be configured independently so that this function
+        // can be used in other generators that wish to generates pairs of spot blocks
+        // with specific constraints.
+        let base_root_epoch_position =
+            gen_spot_root_epoch_position(DEFAULT_SLOT_FILL_RATE, DEFAULT_SLOT_FILL_RATE_DELTA);
+
+        // Constraining the second state to have a greater blockchain length than the
+        // first, we need to constrain the first blockchain length such that there is some room
+        // leftover in the epoch for at least 1 more block to be generated.
+        let _blockchain_length_relativity = Some(&a);
+        let max_epoch_slot = match _blockchain_length_relativity {
+            Some(_blockchain_length_relativity) => {
+                a.body.consensus_state.curr_global_slot.slots_per_epoch.0 - 4
+            } // -1 to bring into inclusive range, -3 to provide 2 slots of fudge room
+            None => a.body.consensus_state.curr_global_slot.slots_per_epoch.0 - 1, // -1 to bring into inclusive range
+        };
```

Comment:
when would the `None` case ever be hit here? since `_blockchain_length_relativity = Some(&a)` it looks like it's always `Some`

Reply:
We use new a, b state then directly use it as _blockchain_length_relativity  and there's never [blockchain_length_relativity](https://github.com/MinaProtocol/mina/blob/af76cb7980d5e81e704120290a850ea9c6f8522e/src/lib/consensus/proof_of_stake.ml#L4258) input from origin mina OCaml just a, b state see [url](https://github.com/MinaProtocol/mina/blob/af76cb7980d5e81e704120290a850ea9c6f8522e/src/lib/consensus/proof_of_stake.ml#L4414)

Reply:
`Ascending and `Equal are just enum variants they define in the code. See https://ocaml.org/learn/tutorials/labels.html "More variants" section at the bottom

Reply:
so the value of blockchain_length_relativity is not `&mut ProtocolState`, but rather either `Ascending or `Equal based on the ocaml code.

Reply:
I have no idea how they decide what is Ascending or Equal in Mina Ocaml. And there's no blockchain_relativity input from Mina Ocaml. Should it means default will be None?

Reply:
> The blockchain_relativety is a state please see the above links I attached

I think you mean `blockchain_length_relativity` and not `blockchain_relativity` as there's no `blockchain_relativity` in code.
ok, let me help see it better. If you take a look at https://github.com/MinaProtocol/mina/blob/af76cb7980d5e81e704120290a850ea9c6f8522e/src/lib/consensus/proof_of_stake.ml#L4140 clearly `blockchain_length_relativity` is being `match`ed with the `Ascending` and `Equal` variant they have defined. If `blockchain_length_relativity` were a `ProtocolState`, then the match arm would have matched with either an instance of protocol state or some of its fields, which isn't the case. hope that helps :)
 

>  I have no idea how they decide what is Ascending or Equal

Those are just enum variants they define, as i replied above. Asceding or Equal is totally different than > or = and they are not at all related here. They are a new custom enum type they have defined.

Reply:
I already repeat the same thing. Hope you don't repeat what I'm saying again, thank you.

Reply:
We use default a, b state then directly use a state as _blockchain_length_relativity root base to create relativity for max_epoch_slot because there's no blockchain_length_relativity in the input argument just a, b state in the second screenshot.
![Screenshot (424)](https://user-images.githubusercontent.com/8589224/150298558-651f2056-8b33-46b4-b346-06515ea9d48f.png)
![Screenshot (425)](https://user-images.githubusercontent.com/8589224/150298572-f5e44976-6cbb-4eb3-a618-58548bcc74e8.png)
.

Reply:
They only set `blockchain_length_relativity` in the selection case which is not include in the is_short_range test case.

Reply:
It should always be none in this is_short_range test cases PR. No need to handle `_blockchain_length_relativity`